<a class="anchor" id="section-0"></a>
<img src="https://i.imgur.com/lTCPbz0.jpeg" alt='logo puc-sp' width='30%' align='left'>

### Pontifícia Universidade Católica de São Paulo (PUC-SP)

<h1 style=font-size:30px>Machine Learning: Desvendando o Fifa 22</h1>


### Bacharelado em Ciência de Dados e Inteligência Artificial

#### Turma: CDIA21-MA

**Professor:** Jefferson de Oliveira Silva

### Integrantes

<table align="left" style=font-size:15px>
<thead>
  <tr>
    <th style="text-align:left;">Nome</th>
    <th style="text-align:left;">RA</th>
    <th></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td style="text-align:left;">Gustavo Schlieper Tessitore</td>
    <td>RA00297844</td>
    <td></td>
  </tr>
  <tr>
    <td style="text-align:left;">Isaac Higuchi</td>
    <td>RA00306191</td>
    <td></td>
  </tr>
  <tr>
    <td style="text-align:left;">João Pedro Taves Araujo</td>
    <td>RA00297753</td>
    <td></td>
  </tr>
  <tr>
    <td style="text-align:left;">Lucas Lopes Amorim</td>
    <td>RA00303799</td>
    <td></td>
  </tr>
</tbody>
</table>

<h3><b>Estrutura:</b></h3>

<ol>
  <li><a href="#section-1">Contextualização</a></li>
  <li><a href="#section-2">Problema de Negócio</a></li>
  <li><a href="#section-3">Método</a>
      <ol style="list-style: none; padding-left: 10px">
          <li>3.1. <a href="#section-3-1">Entendimento dos Dados</a></li>
          <li>3.2. <a href="#section-3-2">Preparação dos Dados</a></li>
              <ol style="list-style: none; padding-left: 20px">
                  <li>3.2.1 <a href="#section-3-2-1">Gerando novos atributos</a></li>
                  <li>3.2.2 <a href="#section-3-2-2">Tratando variáveis categórias</a></li>
              </ol>
          <li>3.3. <a href="#section-3-3">Modelagem</a>
              <ol style="list-style: none; padding-left: 20px">
                  <li>3.3.1 <a href="#section-3-3-1">Modelo #2: Floresta Aleatória</a></li>
                  <li>3.3.2 <a href="#section-3-3-2">Modelo #1: Regressão Linear</a></li>
                  <li>3.3.3 <a href="#section-3-3-3">Modelo #3: Regressão Logística</a></li>
                  <li>3.3.4 <a href="#section-3-3-4">Modelo #4: Máquina de Vetores de Suporte</a></li>
              </ol>
          </li>
          <li>3.4. <a href="#section-3-4">Avaliação dos Modelos</a>
              <ol style="list-style: none; padding-left: 20px">
                  <li>3.4.1 <a href="#section-3-4-1">Modelo #1: Regressão Linear</a></li>
                  <li>3.4.2 <a href="#section-3-4-2">Modelo #2: Floresta Aleatória</a></li>
                  <li>3.4.3 <a href="#section-3-4-3">Modelo #3: Regressão Logística</a></li>
                  <li>3.4.4 <a href="#section-3-4-4">Modelo #4: Máquina de Vetores de Suporte</a></li>
              </ol>
          </li>
          <li>3.5. <a href="#section-3-5">Persistência e Disponibilização dos Modelos</a></li>
    </ol>
  </li>
    <li><a href="#section-4">Resultados</a></li>
  <li href="#section-5"><a href="#section-5">Limitações</a></li>
  <li><a href="#section-6">Conclusões</a></li>
</ol>
<p><a href="#section-n">Referências</a></p>

# 1. Contextualização <a class="anchor" id="section-1"></a>
[Voltar ao topo](#section-0)
![Banner do jogo Fifa 22](https://i.imgur.com/8ATGwt4.jpg)

# TODO: enfatizar que o mercado de e-sports necessita de mais decisões data-driven

Sendo o maior segmento da indústria do entretenimento desde 2003, segundo um infográfico disponibilizado pela En.Digital (com dados de empresas como: Newzoo, Statista etc.) o mercado mundial de games movimentou, apenas em 2021, mais de US\\$175 bilhões, de acordo com dados da consultoria Newzoo, e tem uma projeção até 2023 de crescimento que deve movimentar aproximadamente US\\$200 bilhões.

Dentro desse enorme mercado, que atualmente representa mais de 65% do faturamento da indústria de entretenimento e ultrapassa facilmente o mercado cinematográfico e da música, temos grandes franquias consolidadas e conhecidas globalmente, como por exemplo: Grand Theft Auto, Pokémon, Super Mario, Call of Duty e muitas outras. 

Mas todo ano tem suas certezas: tem Natal em dezembro, quatro estações do ano e o lançamento de um novo FIFA. Todos os anos um novo FIFA nos surpreende faturando muito e reforçando sua marca na história dos games. Seu último lançamento em outubro de 2021, o FIFA 22, já contava com mais de 9 milhões de jogadores e 460 milhões de partidas jogadas nos quatro dias de acesso antecipado para quem comprou o jogo em pré-lançamento, segundo matéria do Canaltech.

Sem dúvida um dos maiores reanimadores dos jogadores de FIFA foi o lançamento do FIFA Ultimate Team (FUT) dentro do jogo. Um modo onde você pode construir seu próprio time e se sentir como dono de um clube. Segundo o Globo Esporte, o modo Ultimate Team de jogos como FIFA 21, Madden NFL 21 e NHL 21 gerou US\\$ 1,62 bilhão (R\\$ 8,4 bilhões) de receita para a Electronic Arts no ano fiscal de 2020 (de abril de 2020 a março de 2021). Se tratava da maior arrecadação com o modo na história da desenvolvedora.

# 2. Problema de Negócio <a class="anchor" id="section-2"></a>
[Voltar ao topo](#section-0)
![Workflow simplificado do projeto](https://i.imgur.com/gjucTUv.png)

O mercado de *E-sports* é muito competitivo e necessita que cada decisão seja tomada cuidadosamente. Para o apoio da tomada de decisão, é preciso analisar uma quantidade muito grande de atributos, o que torna difícil para um ser humano, mesmo sendo um jogador profissional, assimilar todas as "regras de negócio" do jogo para que ele possa manifestar o seu melhor desempenho.

O modo *Ultimate Team* é particularmente mais desafiador. Nele o jogador inicia o jogo com um time modesto e precisa além de desempenhar bem nos jogos, de uma grande capacidade de gerenciamento do seu time para poder evoluir, além de conhecimento sobre futebol porque nem todo jogador bom se encaixa no time por uma questão de entrosamento, seu estilo de jogo e seu gosto pessoal.

# TODO: Mudar o foco de jogador iniciante para jogadores profissionais
# TODO: Destacar como é difícil aprender todas as regras do jogo pela sua complexidade (por haver muitos atributos relacionados a um jogador)

Grandes sucessos geram grandes oportunidades de negócio. Sendo o FIFA, sem qualquer incerteza, uma das maiores franquias de games que existiram do mundo, criar algo com relação à franquia gera, no mínimo, grande atenção do público do jogo. Porém, jogar o modo Ultimate Team demanda habilidade de gerenciamento, já que os melhores jogadores custam valores altos que te “limitam” durante um tempo a ter muitos jogadores famosos e habilidosos, e conhecimento sobre o jogo e futebol porque nem todo jogador bom se encaixa no time devido ao entrosamento do time, seu estilo de jogo e seu gosto pessoal. Todas essas barreiras desanimam um jogador iniciante, logo, a fim de aumentar o público do jogo tornando-o mais acessível à amantes do futebol e engajando novatos no game, decidimos desenvolver um sistema de recomendação de cartas (cartas representam os jogadores de futebol dentro do jogo, contendo informações como: sua foto, nacionalidade, clube atual, overall e principais atributos) onde o jogador do FIFA receberá a recomendação das melhores cartas dada a posição escolhida, seu saldo disponível, 

# 3. Método <a class="anchor" id="section-3"></a>
[Voltar ao topo](#section-0)

## 3.1 Entendimento dos Dados <a class="anchor" id="section-3-1"></a>
[Voltar ao topo](#section-0)

## 3.2 Preparação dos Dados <a class="anchor" id="section-3-2"></a>
[Voltar ao topo](#section-0)

### 3.2.1 Gerando novos atributos <a class="anchor" id="section-3-2-1"></a>
[Voltar ao topo](#section-0)

Os dados do nosso projeto foram armazenados no BigQuery, para poder consulta-los criamos uma função `execute` que pode ser encontrada em `fifa22-players-analysis/database/bq_connection.py`.
Executamos a consulta:
```sql
SELECT * FROM `puc-sp.fifa.player`
```
E salvamos os dados em um arquivo tsv `raw_data.tsv`

In [18]:
import pandas as pd
data = pd.read_csv("./preprocessing/data/raw_data.tsv", sep="\t", low_memory=False)
data

sofifa_id                                         player_url  \
0         167575  https://sofifa.com/player/167575/anton-shunin/...   
1         205211  https://sofifa.com/player/205211/milan-borjan/...   
2         244700  https://sofifa.com/player/244700/yuriy-dyupin/...   
3         199904  https://sofifa.com/player/199904/jesse-joronen...   
4         240225  https://sofifa.com/player/240225/matvey-safono...   
...          ...                                                ...   
19234     262593  https://sofifa.com/player/262593/moritz-seiffe...   
19235     262614  https://sofifa.com/player/262614/till-muschkow...   
19236     262755  https://sofifa.com/player/262755/shalva-ogbaid...   
19237     262756  https://sofifa.com/player/262756/matteo-gumane...   
19238     263236  https://sofifa.com/player/263236/yazid-heimur/...   

           short_name            long_name player_positions  overall  \
0           A. Shunin         Anton Shunin               GK       76   
1           M. Borjan         Milan Borjan               GK       76   
2           Y. Dyupin         Yuriy Dyupin               GK       74   
3          J. Joronen  Jesse Pekka Joronen               GK       72   
4          M. Safonov       Matvey Safonov               GK       72   
...               ...                  ...              ...      ...   
19234     M. Seiffert      Moritz Seiffert               LW       58   
19235  T. Muschkowski     Till Muschkowski               RW       55   
19236     S. Ogbaidze      Shalva Ogbaidze               ST       55   
19237      M. Gumaneh       Matteo Gumaneh               ST       55   
19238       Y. Heimur         Yazid Heimur               CM       55   

       potential  value_eur  wage_eur         dob  ...   cdm   rdm   rwb  \
0             76        NaN       NaN  1987-01-27  ...  29+2  29+2  27+2   
1             76        NaN       NaN  1987-10-23  ...  37+2  37+2  31+2   
2             74        NaN       NaN  1988-03-17  ...  22+2  22+2  19+2   
3             72        NaN       NaN  1993-03-21  ...  27+2  27+2  26+2   
4             82        NaN       NaN  1999-02-25  ...  24+2  24+2  19+2   
...          ...        ...       ...         ...  ...   ...   ...   ...   
19234         66   450000.0     600.0  2000-11-04  ...  40+2  40+2  44+2   
19235         65   275000.0     550.0  2001-05-06  ...  44+2  44+2  47+2   
19236         65   275000.0     500.0  2002-08-01  ...  39+2  39+2  40+2   
19237         67   300000.0     500.0  2002-09-05  ...  37+2  37+2  40+2   
19238         66   275000.0     500.0  2002-11-18  ...  51+2  51+2  48+2   

         lb   lcb    cb   rcb    rb    gk nationality_name  
0      27+2  28+2  28+2  28+2  27+2  75+1           Russia  
1      31+2  35+2  35+2  35+2  31+2  75+1           Canada  
2      18+2  22+2  22+2  22+2  18+2  73+1           Russia  
3      26+2  27+2  27+2  27+2  26+2  71+1          Finland  
4      19+2  25+2  25+2  25+2  19+2  71+2           Russia  
...     ...   ...   ...   ...   ...   ...              ...  
19234  42+2  37+2  37+2  37+2  42+2  14+2          Germany  
19235  45+2  39+2  39+2  39+2  45+2  11+2          Germany  
19236  38+2  36+2  36+2  36+2  38+2  14+2          Georgia  
19237  38+2  34+2  34+2  34+2  38+2  15+2          Germany  
19238  48+2  48+2  48+2  48+2  48+2  16+2          Germany  

[19239 rows x 91 columns]

Nesse conjunto de dados precisamos fazer algumas tratativas para poder passar o máximo de informação possível para os modelos.
Uma informação que é de grande interesse é a idade dos jogadores, mas temos apenas a coluna `dob` (date of birth), mas podemos facilmente calcular a idade.

In [19]:
from datetime import date, datetime

current_date = datetime.today()
data.dob = pd.to_datetime(data.dob)

def age(dob):
    today = date.today()
    return today.year - dob.year - ((today.month, today.day) < (dob.month, dob.day))

data["age"] = data.dob.apply(age)
data[["dob", "age"]].head()

dob  age
0 1987-01-27   35
1 1987-10-23   34
2 1988-03-17   34
3 1993-03-21   29
4 1999-02-25   23

Outra característica dos jogadores que é do nosso interesse é o indice de massa corporal,
e alguns atributos falando a respeito do grau de obesidade ou magreza dos jogadores.
De acordo com o site [saúde fácil](https://www.programasaudefacil.com.br/calculadora-de-imc) temos a seguinte tabela
para classificar o imc:

<br>
<table align="left" style=font-size:15px>
<thead>
  <tr>
    <th style="text-align:left;">IMC</th>
    <th style="text-align:left;">Classificação</th>
    <th></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td style="text-align:left;">MENOR QUE 18,5</td>
    <td>MAGREZA</td>
    <td></td>
  </tr>
  <tr>
    <td style="text-align:left;">ENTRE 18,5 E 24,9</td>
    <td>NORMAL</td>
    <td></td>
  </tr>
  <tr>
    <td style="text-align:left;">ENTRE 25,0 E 29,9</td>
    <td>SOBREPESO</td>
    <td></td>
  </tr>
  <tr>
    <td style="text-align:left;">ENTRE 30,0 E 39,9</td>
    <td>OBESIDADE</td>
    <td></td>
  </tr>
</tbody>
</table>

Agora para poder implementar essas informações no nosso conjunto de dados vamos
precisar da ajuda dos conceitos aprendidos nas aulas de Engenharia de Software.
Primeiramente precisamos definir a fórmula do IMC, que é descrita pela seguinte função:

$$IMC = \frac{peso}{altura^2}$$

O peso é em Kg e a altura em metros.

Agora precisamos definir como essas informações seriam implementada no mundo lógico para depois traduzi-la para Python. Para isso vamos precisar das seguintes variáveis:
<br>
VARIÁVEIS
$$peso: Peso$$
$$altura: Altura$$
$$imc: Float$$
$$obesidade: Boolean$$
$$magreza: Boolean$$
$$normal: Boolean$$
$$válido: Boolean$$
$$jogadores: Sequencia$$

<br>

Assim temos que o nosso predicado de inicio é:

$$inicio \triangleq\wedge peso = \bot$$
$$\wedge altura = \bot$$
$$\wedge imc = \bot$$
$$\wedge obesidade = \bot$$
$$\wedge magreza = \bot$$
$$\wedge normal = \bot$$
$$\wedge válido = False$$
$$\wedge jogadores = \langle ~ \rangle$$

<br>

Para ajudar na implementação vamos definir os tipos Peso e Altura, mas antes vamos entender os limites desses tipos

In [33]:
max_weight = data.weight_kg.max()
min_weight = data.weight_kg.min()
print(f"Peso entre {min_weight:.2f} Kg e {max_weight:.2f} Kg")

max_height = data.height_cm.max() / 100
min_height = data.height_cm.min() / 100
print(f"Altura entre {min_height:.2f} m e {max_height:.2f} m")

Peso entre 49.00 Kg e 110.00 Kg
Altura entre 1.55 m e 2.06 m


$$Peso \triangleq [~p: Float~|~40 < p < 150~]$$
$$Altura \triangleq [~a: Float~|~1.4 < a < 2.2~]$$

<br>

Assim definimos os tipos $Peso$ e $Altura$ com os limites conforme o nosso conjunto de dados mais uma margem de erro arbitrária

<br>

Agora vamos definir a função que calcula o IMC:

$$calculaIMC \triangleq \lambda peso : Peso, ~ altura : Altura \bullet \frac{peso}{altura^2}$$

Uma que verifica a obesidade:

$$verificaObesidade \triangleq \lambda imc : Float \bullet ~\textbf{if}~ imc \geq 30 ~\textbf{then}~ True ~\textbf{else}~ False$$

Uma que verifica a magreza:

$$verificaMagreza \triangleq \lambda imc : Float \bullet ~\textbf{if}~ imc \leq 18.5 ~\textbf{then}~ True ~\textbf{else}~ False$$

E outra que verifica a normalidade:

$$verificaNormal \triangleq \lambda imc : Float \bullet ~\textbf{if}~ 18.5 < imc < 30 ~\textbf{then}~ True ~\textbf{else}~ False$$

<br>

Classificando um jogador

$$classificaJogador_1[p,~a] \triangleq\wedge Peso(p)~\wedge Altura(a)$$
$$\wedge peso' = Peso(p)$$
$$\wedge altura' = Altura(a)$$
$$\wedge imc' = calculaIMC(peso', altura')$$
$$\wedge obesidade' = verificaObesidade(imc')$$
$$\wedge magreza' = verificaMagreza(imc')$$
$$\wedge normal' = verificaNormal(imc')$$
$$\wedge jogadores' = jogadores \frown \langle(peso', altura', imc', obesidade', magreza')\rangle$$
$$\wedge válido' = True$$
<br>
<br>
$$classificaJogador_2[p,~a] \triangleq \wedge (\lnot Peso(p) \lor \lnot Altura(a))$$
$$\wedge peso' = \bot$$
$$\wedge altura' = \bot$$
$$\wedge imc' = \bot$$
$$\wedge obesidade' = \bot$$
$$\wedge magreza' = \bot$$
$$\wedge normal' = \bot$$
$$\wedge jogadores' = jogadores$$
$$\wedge válido' = False$$
<br>
<br>
$$classificaJogador[p,~a] \triangleq classificaJogador_1[p,~a] \lor classificaJogador_2[p,~a]$$

Agora implementando essa lógica em python

In [37]:
# Inicio
peso: float = 0
altura: float = 0
imc: float = 0
obesidade: bool = False
magreza: bool = False
normal: bool = False
valido: bool = False
jogadores: list = []
# Adicionar válido



def valida_peso(p):
    return 40 < p < 150


def valida_altura(a):
    return 1.4 < a < 2.2


def calcula_imc(peso: float, altura: float) -> float:
    return peso / (altura ** 2)


def verifica_obesidade(imc: float) -> bool:
    return imc >= 30


def verifica_magreza(imc: float) -> bool:
    return imc <= 18.5


def verifica_normal(imc: float) -> bool:
    return 18.5 < imc < 30


def classifica_jogador(peso: float, altura: float) -> None:
    global jogadores
    if valida_altura(altura) and valida_peso(peso):
        imc = calcula_imc(peso, altura)
        obesidade = verifica_obesidade(imc)
        magreza = verifica_magreza(imc)
        normal = verifica_normal(imc)
        valido = True
    else:
        raise ValueError("Peso ou altura inválidos")
    jogadores.append((imc, obesidade, magreza, normal, valido))


# Iterando sobre o dataframe
for row in data.itertuples():
    classifica_jogador(row.weight_kg, row.height_cm / 100)

jogadores_df = pd.DataFrame(jogadores, columns=["imc", "obesidade", "magreza", "normal", "valido"])
jogadores_df.head()

imc  obesidade  magreza  normal  valido
0  24.122146      False    False    True    True
1  21.865889      False    False    True    True
2  24.021276      False    False    True    True
3  22.932825      False    False    True    True
4  22.243924      False    False    True    True

In [39]:
data["BMI"] = jogadores_df.imc
data["obese"] = jogadores_df.obesidade.astype(int)
data["lean"] = jogadores_df.magreza.astype(int)
data["normal"] = jogadores_df.normal.astype(int)
data.head()

sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name player_positions  overall  potential  value_eur  \
0         Anton Shunin               GK       76         76        NaN   
1         Milan Borjan               GK       76         76        NaN   
2         Yuriy Dyupin               GK       74         74        NaN   
3  Jesse Pekka Joronen               GK       72         72        NaN   
4       Matvey Safonov               GK       72         82        NaN   

   wage_eur        dob  ...    cb   rcb    rb    gk  nationality_name age  \
0       NaN 1987-01-27  ...  28+2  28+2  27+2  75+1            Russia  35   
1       NaN 1987-10-23  ...  35+2  35+2  31+2  75+1            Canada  34   
2       NaN 1988-03-17  ...  22+2  22+2  18+2  73+1            Russia  34   
3       NaN 1993-03-21  ...  27+2  27+2  26+2  71+1           Finland  29   
4       NaN 1999-02-25  ...  25+2  25+2  19+2  71+2            Russia  23   

         BMI  obese  lean normal  
0  24.122146      0     0      1  
1  21.865889      0     0      1  
2  24.021276      0     0      1  
3  22.932825      0     0      1  
4  22.243924      0     0      1  

[5 rows x 96 columns]

### 3.2.2 Tratando variáveis categórias <a class="anchor" id="section-3-2-2"></a>
[Voltar ao topo](#section-0)

Outra tratativa a ser feita é transformar cada `player_traits` e cada `player_tags` diferente numa coluna boleana
Essa lógica será aplicada em `player_positions`, `work_rate`, `body_type`, `preferred_foot` e `nationality_name`

In [53]:
data[["player_traits", "player_tags", "player_positions", "work_rate", "body_type", "preferred_foot", "nationality_name"]]

player_traits player_tags player_positions  \
0                               Team Player         NaN               GK   
1             Leadership, Comes For Crosses         NaN               GK   
2                                       NaN         NaN               GK   
3      GK Long Throw, Cautious With Crosses         NaN               GK   
4                                       NaN         NaN               GK   
...                                     ...         ...              ...   
19234                                   NaN         NaN               LW   
19235                                   NaN         NaN               RW   
19236                                   NaN         NaN               ST   
19237                                   NaN         NaN               ST   
19238                                   NaN         NaN               CM   

           work_rate       body_type preferred_foot nationality_name  
0      Medium/Medium     Lean (185+)           Left           Russia  
1      Medium/Medium   Normal (185+)          Right           Canada  
2      Medium/Medium   Normal (185+)          Right           Russia  
3      Medium/Medium   Normal (185+)          Right          Finland  
4      Medium/Medium   Normal (185+)          Right           Russia  
...              ...             ...            ...              ...  
19234  Medium/Medium  Lean (170-185)           Left          Germany  
19235  Medium/Medium  Lean (170-185)          Right          Germany  
19236  Medium/Medium   Normal (185+)          Right          Georgia  
19237  Medium/Medium   Normal (185+)          Right          Germany  
19238  Medium/Medium   Normal (185+)          Right          Germany  

[19239 rows x 7 columns]

In [44]:
# Removendo caracteres especiais
data["player_tags"].replace("\xa0", "", regex=True, inplace=True)

# Convertendo variáveis categoricas nominais em dummies

player_traits_dummies = data.player_traits.str.get_dummies(sep=', ')
player_traits_dummies.columns = ['trait '+col_name for col_name in player_traits_dummies.columns]

player_tags_dummies = data.player_tags.str.get_dummies(sep=', ')
player_tags_dummies.columns = ['tag '+col_name for col_name in player_tags_dummies.columns]

player_positions_dummies = data.player_positions.str.get_dummies(sep=', ')

work_rate_dummies = data.work_rate.str.get_dummies()
work_rate_dummies.columns = ['work rate ' + col_name for col_name in work_rate_dummies.columns]

body_type_dummies = data.body_type.str.get_dummies()
body_type_dummies.columns = ['body type ' + col_name for col_name in body_type_dummies.columns]

preferred_foot_dummies = data.preferred_foot.str.get_dummies()
preferred_foot_dummies.columns = ['preferred foot ' + col_name for col_name in preferred_foot_dummies.columns]

nationality_name_dummies = data.nationality_name.str.get_dummies()
nationality_name_dummies.columns = ['nationality ' + col_name for col_name in nationality_name_dummies.columns]

processed_data = pd.concat(
    [
        data, player_traits_dummies, player_tags_dummies,
        player_positions_dummies, work_rate_dummies, body_type_dummies,
        preferred_foot_dummies, nationality_name_dummies
    ], axis=1
)

processed_data.head()

sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name player_positions  overall  potential  value_eur  \
0         Anton Shunin               GK       76         76        NaN   
1         Milan Borjan               GK       76         76        NaN   
2         Yuriy Dyupin               GK       74         74        NaN   
3  Jesse Pekka Joronen               GK       72         72        NaN   
4       Matvey Safonov               GK       72         82        NaN   

   wage_eur        dob  ...  nationality Ukraine  \
0       NaN 1987-01-27  ...                    0   
1       NaN 1987-10-23  ...                    0   
2       NaN 1988-03-17  ...                    0   
3       NaN 1993-03-21  ...                    0   
4       NaN 1999-02-25  ...                    0   

   nationality United Arab Emirates  nationality United States  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3                                 0                          0   
4                                 0                          0   

  nationality Uruguay  nationality Uzbekistan nationality Venezuela  \
0                   0                       0                     0   
1                   0                       0                     0   
2                   0                       0                     0   
3                   0                       0                     0   
4                   0                       0                     0   

   nationality Vietnam  nationality Wales  nationality Zambia  \
0                    0                  0                   0   
1                    0                  0                   0   
2                    0                  0                   0   
3                    0                  0                   0   
4                    0                  0                   0   

  nationality Zimbabwe  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  

[5 rows x 337 columns]

In [45]:
# Convertendo o tipo das colunas
columns_to_evaluate = "ls,st,rs,lw,lf,cf,rf,rw,lam,cam,ram,lm,lcm,cm,rcm,rm,lwb,ldm,cdm,rdm,rwb,lb,lcb,cb,rcb,rb,gk".split(",")
for column_to_evaluate in columns_to_evaluate:
    evaluated_column = processed_data[column_to_evaluate].apply(lambda cell: eval(cell))
    processed_data[column_to_evaluate] = evaluated_column

processed_data.head()

sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name player_positions  overall  potential  value_eur  \
0         Anton Shunin               GK       76         76        NaN   
1         Milan Borjan               GK       76         76        NaN   
2         Yuriy Dyupin               GK       74         74        NaN   
3  Jesse Pekka Joronen               GK       72         72        NaN   
4       Matvey Safonov               GK       72         82        NaN   

   wage_eur        dob  ...  nationality Ukraine  \
0       NaN 1987-01-27  ...                    0   
1       NaN 1987-10-23  ...                    0   
2       NaN 1988-03-17  ...                    0   
3       NaN 1993-03-21  ...                    0   
4       NaN 1999-02-25  ...                    0   

   nationality United Arab Emirates  nationality United States  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3                                 0                          0   
4                                 0                          0   

  nationality Uruguay  nationality Uzbekistan nationality Venezuela  \
0                   0                       0                     0   
1                   0                       0                     0   
2                   0                       0                     0   
3                   0                       0                     0   
4                   0                       0                     0   

   nationality Vietnam  nationality Wales  nationality Zambia  \
0                    0                  0                   0   
1                    0                  0                   0   
2                    0                  0                   0   
3                    0                  0                   0   
4                    0                  0                   0   

  nationality Zimbabwe  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0  

[5 rows x 337 columns]

In [46]:
import os

# Função auxiliar para salvar os dados
def get_project_root_path():
    project_directory = "fifa22-players-analysis"
    separated_path = os.getcwd().split(os.sep)
    project_directory_index = separated_path.index(project_directory) + 1
    return os.sep.join(separated_path[0:project_directory_index])

data_path = os.sep.join([get_project_root_path(), "preprocessing", "data", f"processed_data.tsv"])
processed_data.to_csv(data_path, index=False, sep="\t")

### 3.2.3 Tratando valores nulos <a class="anchor" id="section-3-2-3"></a>
[Voltar ao topo](#section-0)

In [47]:
# Função auxiliar que conta valores nulos
def count_na(df: pd.DataFrame) -> None:
    nan_sum = df.isna().sum()
    display(nan_sum[nan_sum > 0])

count_na(data)

value_eur                74
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
player_tags           17798
player_traits          9841
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

Analisando essa tabela com os valores nulos podemos perceber que alguns jogadores do dataset não possuem um time atual, isso explica
`wage_eur`, `club_team_id`, `club_position`, `club_jersey_number` terem a mesma quantidade de valores nulos (61),
porém `value_eur` deveria seguir esse mesmo padrão, assim podemos dizer que 13 (74 - 61) jogadores que estão ativos não possuem um valor,
portanto precisamos preencher esses dados.

Também podemos perceber que temos 2132 jogadores não possuem `pace`, `shooting`, `passing`, `dribbling`, `defending`, `physic`. Vamos analisar esses jogadores para entender quem são.

In [52]:
data = processed_data
def percent_of_gk(df: pd.DataFrame) -> None:
    gk = df[df['GK'] == 1]
    print(f"Porcentagem de goleiros sem atributos principais {gk.shape[0] / df.shape[0] * 100}%")

missing_main_attr_data = data[data.pace.isna()]
percent_of_gk(missing_main_attr_data)

Porcentagem de goleiros sem atributos principais 100.0%


Ou seja, no nosso dataset os goleiros não têm os atributos principais, portanto vamos preencher-los também

In [54]:
missing_gk_speed_data = data[data.goalkeeping_speed.isna()]
sum_of_missing_data = missing_gk_speed_data.shape[0] + missing_main_attr_data.shape[0]
print(f"Jogadores sem gk speed + jogadores sem atributos primários: {sum_of_missing_data}")
print(f"Numero de jogadores: {data.shape[0]}")

Jogadores sem gk speed + jogadores sem atributos primários: 19239
Numero de jogadores: 19239


Com isso podemos afirmar que jogadores que não são goleiros não possuem `goalkeeping_speed`. Tabém vamos prever esses valores.

## 3.3 Modelagem <a class="anchor" id="section-3-3"></a>
[Voltar ao topo](#section-0)

A metodologia usada nesse processo é formada pelas seguintes etapas:
- Seprara features e target
- Preencher valores nulos que com a média respectiva de cada coluna
- Separar dadados de treino e teste
- Salvar o modelo treinado (opcional)
- Calcular métricas de avaliação
- Prever valores nulos do dataset a partir do modelo treinado
- Adicionar esses valores ao conjunto de dados original

O código a seguir representa a implementação da metodologia descrita:

In [75]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
import numpy as np
import joblib


def save_model(model, model_name):
    filename = f"model_name_{model_name}.pkl"
    print(filename)
    model_path = os.sep.join([get_project_root_path(), "saved_models", filename])
    joblib.dump(model, model_path)


def predict_and_populate(df_to_populate, df_to_train, model, column_to_predict, round_=None, save=False) -> pd.DataFrame:
    # Separating the target variable
    target = df_to_train[column_to_predict]
    features = df_to_train.drop(column_to_predict, axis=1)

    # Poluting the missing features
    mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
    mean_imputer.fit(features)
    imputed_features = mean_imputer.transform(features)

    # Train test split
    features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.3)

    # Training the model
    trained_model = model.fit(features_train, target_train)

    # Saving the model
    if save:
        model_name = str(model).split("(")[0]
        filename = f"{model_name}_{column_to_predict}"
        save_model(trained_model, filename)

    # Calculating model score
    model_score = trained_model.score(features_test, target_test)
    mse = mean_squared_error(target_test, trained_model.predict(features_test))
    mae = mean_absolute_error(target_test, trained_model.predict(features_test))

    print(f"Score do R² modelo: {model_score:.4f}")
    print(f"MSE: {mse:.4f}")
    print(f"MAE: {mae:.4f}")

    # Populating the missing values
    feature_to_populate = mean_imputer.transform(df_to_populate.drop(column_to_predict, axis=1))
    predicted_values = trained_model.predict(feature_to_populate)


    if round_ is not None:
        predicted_values = predicted_values.round(round_)

    df_to_populate[column_to_predict] = predicted_values

    return df_to_populate

### 3.3.1 Modelo #2: Floresta Aleatória <a class="anchor" id="section-3-3-1"></a>
[Voltar ao topo](#section-0)

### Prevendo valor dos jogadores
Nessa etapa vamos criar um modelo para prever o valor (`value_eur`) dos 13 jogadores que não possuem valor.
Primeiros vamos separar esses jogadores e remover as colunas que não são de nosso interesse.

In [56]:
players_with_club_and_no_value = data.query("not club_team_id.isna() and value_eur.isna()").copy()
data_without_players_with_club_and_no_value = data.drop(players_with_club_and_no_value.index)

data_to_train = data_without_players_with_club_and_no_value.query("not value_eur.isna()").copy()

columns_to_drop = [
    "sofifa_id", "player_url", "short_name", "long_name",
    "dob", "club_position","player_traits", "player_tags",
    "player_positions", "work_rate", "body_type",
    "preferred_foot", "nationality_name"
]

data_to_train = data_to_train.drop(columns=columns_to_drop)
data_to_fill = players_with_club_and_no_value.drop(columns=columns_to_drop)

Agora que temos os nossos datasets para treino e para popular podemos comparar alguns modelos diferentes para escolher o melhor.
Comparando `DummyRegressor` vs `LinearRegression` vs `RandomForestRegressor` para prever `value_eur`

In [69]:
from sklearn.dummy import DummyRegressor
dummy = DummyRegressor()
predict_and_populate(data_to_fill, data_to_train, dummy, "value_eur");

Score do R² modelo: -0.0000
MSE: 66771834595772.8906
MAE: 3173503.9321


In [70]:
from sklearn.linear_model import LinearRegression

linear_regression = LinearRegression()
predict_and_populate(data_to_fill, data_to_train, linear_regression, "value_eur");

Score do R² modelo: 0.9742
MSE: 1431395678792.1587
MAE: 483299.0781


In [83]:
from sklearn.ensemble import RandomForestRegressor

random_forest = RandomForestRegressor(random_state=0, n_jobs=-1, oob_score=True)
players_with_club_and_no_value = predict_and_populate(data_to_fill, data_to_train, random_forest, "value_eur", save=True);

model_name_RandomForestRegressor_value_eur.pkl
Score do R² modelo: 0.9928
MSE: 378639144031.4636
MAE: 163750.7644


De cara ja é possivel ver que os modelos de `LinearRegression` e `RandomForestRegressor` são melhores que o modelo base `DummyRegressor`.
Porém, o `RandomForestRegressor` gerou um $R^2$ e um erro absoluto médio menor.


In [84]:
# Populando
data.loc[players_with_club_and_no_value.index, "value_eur"] = players_with_club_and_no_value.value_eur
data.loc[players_with_club_and_no_value.index]

sofifa_id                                         player_url  \
11929     115909  https://sofifa.com/player/115909/ruben-castro-...   
11987     124344  https://sofifa.com/player/124344/ibrahim-oztur...   
12264     153066  https://sofifa.com/player/153066/lucas-licht/2...   
13149     152912  https://sofifa.com/player/152912/jose-sand/220002   
13177     110381  https://sofifa.com/player/110381/maximiliano-r...   
14103     262371  https://sofifa.com/player/262371/craig-dundas/...   
14611     140222  https://sofifa.com/player/140222/paulo-da-silv...   
15125     138703  https://sofifa.com/player/138703/andres-daless...   
16065     158293  https://sofifa.com/player/158293/zhi-zheng/220002   
16498       2702  https://sofifa.com/player/2702/kevin-ellison/2...   
16814      29118  https://sofifa.com/player/29118/shinji-ono/220002   
18159      29552  https://sofifa.com/player/29552/shunsuke-nakam...   
18166     254704  https://sofifa.com/player/254704/kazuyoshi-miu...   

            short_name                     long_name player_positions  \
11929     Rubén Castro           Rubén Castro Martín               ST   
11987        İ. Öztürk                İbrahim Öztürk               CB   
12264         L. Licht            Lucas Matías Licht      LB, LWB, LM   
13149          J. Sand             José Gustavo Sand               ST   
13177     M. Rodríguez   Maximiliano Rubén Rodríguez       LW, LM, ST   
14103        C. Dundas                  Craig Dundas               ST   
14611      P. Da Silva  Paulo César Da Silva Barrios               CB   
15125  A. D'Alessandro   Andrés Nicolás D'Alessandro          CAM, RM   
16065        Zheng Zhi                            郑智          CDM, CM   
16498       K. Ellison           Kevin Keith Ellison           LM, ST   
16814           S. Ono                         小野 伸二          CAM, CM   
18159      S. Nakamura                         中村 俊輔      CAM, CF, RW   
18166         K. Miura                         三浦 知良               ST   

       overall  potential  value_eur  wage_eur        dob  ...  \
11929       69         69   968250.0    3000.0 1981-06-27  ...   
11987       67         67   826500.0    3000.0 1981-06-21  ...   
12264       69         69   933000.0    5000.0 1981-04-06  ...   
13149       75         75  3959250.0   13000.0 1980-07-17  ...   
13177       73         73  2340000.0   10000.0 1981-01-02  ...   
14103       56         56   179750.0    1000.0 1981-02-16  ...   
14611       67         67   822000.0     500.0 1980-02-01  ...   
15125       75         75  4285000.0     900.0 1981-04-15  ...   
16065       67         67   837750.0    5000.0 1980-08-20  ...   
16498       58         58   243300.0    2000.0 1979-02-23  ...   
16814       58         58   244400.0     800.0 1979-09-27  ...   
18159       63         63   552000.0     800.0 1978-06-24  ...   
18166       59         59   323800.0     700.0 1967-02-26  ...   

       nationality Ukraine  nationality United Arab Emirates  \
11929                    0                                 0   
11987                    0                                 0   
12264                    0                                 0   
13149                    0                                 0   
13177                    0                                 0   
14103                    0                                 0   
14611                    0                                 0   
15125                    0                                 0   
16065                    0                                 0   
16498                    0                                 0   
16814                    0                                 0   
18159                    0                                 0   
18166                    0                                 0   

       nationality United States nationality Uruguay  nationality Uzbekistan  \
11929                          0                   0                       0   
11987  

In [85]:
count_na(data)

value_eur                61
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
player_tags           17798
player_traits          9841
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

In [86]:
# Removendo jogadores não ativos
data = data.query("not club_team_id.isna()").copy()

count_na(data)

release_clause_eur     1115
player_tags           17744
player_traits          9823
pace                   2124
shooting               2124
passing                2124
dribbling              2124
defending              2124
physic                 2124
goalkeeping_speed     17054
dtype: int64

### 3.3.2 Modelo #1: Regressão Linear <a class="anchor" id="section-3-3-2"></a>
[Voltar ao topo](#section-0)

### Prevendo pace, shooting, passing, dribbling, defending, physic dos goleiros

In [87]:
gk_data = data.query("GK == 1").copy()
data_without_gk = data.query("GK == 0").copy()

columns_to_predict = ["pace", "shooting", "passing", "dribbling", "defending", "physic"]

for column_to_predict in columns_to_predict:

    columns_to_drop_with_other_atributes = [col for col in columns_to_predict if col != column_to_predict]
    columns_to_drop_with_other_atributes.extend(columns_to_drop)

    data_to_train = data_without_gk.drop(columns=columns_to_drop_with_other_atributes)
    data_to_predict = gk_data.drop(columns=columns_to_drop_with_other_atributes)

    linear_regression = LinearRegression()
    print(f"Prevendo {column_to_predict}")
    filled_data = predict_and_populate(
        data_to_predict, data_to_train, linear_regression,
        column_to_predict, 0, save=True
    )
    print()

    data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]

count_na(data)

Prevendo pace
model_name_LinearRegression_pace.pkl
Score do R² modelo: 0.9993
MSE: 0.0857
MAE: 0.2511

Prevendo shooting
model_name_LinearRegression_shooting.pkl
Score do R² modelo: 0.9990
MSE: 0.1941
MAE: 0.3545

Prevendo passing
model_name_LinearRegression_passing.pkl
Score do R² modelo: 0.9961
MSE: 0.3937
MAE: 0.4799

Prevendo dribbling
model_name_LinearRegression_dribbling.pkl
Score do R² modelo: 0.9991
MSE: 0.0874
MAE: 0.2553

Prevendo defending
model_name_LinearRegression_defending.pkl
Score do R² modelo: 0.9997
MSE: 0.0839
MAE: 0.2508

Prevendo physic
model_name_LinearRegression_physic.pkl
Score do R² modelo: 0.9991
MSE: 0.0845
MAE: 0.2508



release_clause_eur     1115
player_tags           17744
player_traits          9823
goalkeeping_speed     17054
dtype: int64

### Prevendo goalkeeping_speed dos jogadores

In [88]:
players_data = data.query("GK == 0").copy()
data_to_predict = players_data.drop(columns=columns_to_drop)
data_to_train = gk_data.drop(columns=columns_to_drop)

column_to_predict = "goalkeeping_speed"

linear_regression = LinearRegression()
filled_data = predict_and_populate(
    data_to_predict, data_to_train, linear_regression, column_to_predict, 0, save=True
)

data.loc[filled_data.index, column_to_predict] = filled_data[column_to_predict]
count_na(data)

model_name_LinearRegression_goalkeeping_speed.pkl
Score do R² modelo: 0.9993
MSE: 0.0866
MAE: 0.2531


release_clause_eur     1115
player_tags           17744
player_traits          9823
dtype: int64

In [89]:
# Removendo release_clause_eur
data.drop(columns=["release_clause_eur"], inplace=True)

filled_data_save_path = os.sep.join([get_project_root_path(), "preprocessing", "data", f"filled_data.tsv"])
data.to_csv(filled_data_save_path, index=False, sep="\t")

### 3.3.3 Modelo #3: Regressão Logística <a class="anchor" id="section-3-3-3"></a>
[Voltar ao topo](#section-0)

Separando preditores e labels

In [8]:
X, y = features.drop('tag #FK Specialist', axis=1), filtered_df['tag #FK Specialist']

Separando dados de treino e teste

In [9]:
X_train, X_test, y_train, y_test = train_test_split(
    filtered_df[["skill_fk_accuracy"]], filtered_df['tag #FK Specialist'],
    random_state=0
)

Após a nossa minuciosa análise, decidimos usar um modelo de regressão logística com validação cruzada maximizar o *recall* (métrica). 

Tal decisão foi tomada pois o **nosso conjunto de dados é extremamente desbalanceado** (há muito mais verdadeiros negativos do que verdadeiros positivos), e nas configurações padrão do modelo acabamos com uma acurácia alta, porém com uma alta taxa de falsos positivos



In [10]:
# Treinando o modelo
model_reglog = LogisticRegressionCV(scoring='recall').fit(X_train, y_train)

### 3.3.4 Modelo #4: Máquina de Vetores de Suporte <a class="anchor" id="section-3-3-4"></a>
[Voltar ao topo](#section-0)

## 3.4 Avaliação dos Modelos <a class="anchor" id="section-3-4"></a>
[Voltar ao topo](#section-0)

In [11]:
# Para calcular as métricas de performance dos modelos
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import recall_score

In [12]:
# Para gerar visualizações
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set(font_scale=2)

### 3.4.1 Modelo #1: Regressão Linear <a class="anchor" id="section-3-4-1"></a>
[Voltar ao topo](#section-0)

### 3.4.2 Modelo #2: Floresta Aleatória <a class="anchor" id="section-3-4-2"></a>
[Voltar ao topo](#section-0)

### 3.4.3 Modelo #3: Regressão Logística <a class="anchor" id="section-3-4-3"></a>
[Voltar ao topo](#section-0)

Usando nosso modelo para predizer os valores de y do nosso dataset de teste

In [13]:
y_pred = model_reglog.predict(X_test)

**Acurácia**

In [14]:
acuracia = accuracy_score(y_test, y_pred)
print(f'acurácia = {acuracia:.3f}')

acurácia = 1.000


**Recall**

In [15]:
rc = recall_score(y_test, y_pred)
print(f'recall = {rc:.3f}')

recall = 1.000


**Verificando coeficiente de correlação para cada uma das features**

In [16]:
for i in range(len(model.coef_)):
    print(f'model.coef_  = {model.coef_[i].round(3)}')

NameError: name 'model' is not defined

In [ ]:
cf_matrix = confusion_matrix(y_test, y_pred)
group_names = ['True Neg','False Pos','False Neg','True Pos']
group_counts = ["{0:0.0f}".format(value) for value in
                cf_matrix.flatten()]
group_percentages = ["{0:.2%}".format(value) for value in
                     cf_matrix.flatten()/np.sum(cf_matrix)]
labels = [f"{v1}\n{v2}\n{v3}" for v1, v2, v3 in
          zip(group_names,group_counts,group_percentages)]
labels = np.asarray(labels).reshape(2,2)

In [ ]:
plt.figure(figsize=(15,8))
plt.title('Matriz de confusão: Esse jogador é um #FK Specialist?', weight='bold').set_fontsize(30)
sns.heatmap(confusion_matrix(y_test, y_pred), annot=labels, fmt='', cmap='Blues')
plt.xlabel(f'Predicted value\n\naccuracy = {acuracia:.3f}\nrecall = {rc:.3f}')
plt.ylabel('True value')
plt.show()

### 3.4.4 Modelo #4: Máquina de Vetores de Suporte <a class="anchor" id="section-3-4-4"></a>
[Voltar ao topo](#section-0)

## 3.5 Persistência e Disponibilização dos Modelos <a class="anchor" id="section-3-5"></a>
[Voltar ao topo](#section-0)

# 4. Resultados <a class="anchor" id="section-4"></a>
[Voltar ao topo](#section-0)

# 5. Limitações <a class="anchor" id="section-5"></a>
[Voltar ao topo](#section-0)

# 6. Conclusões <a class="anchor" id="section-6"></a>
[Voltar ao topo](#section-0)

# Referências<a class="anchor" id="section-n"></a>
[Voltar ao topo](#section-0)

- [FIFA 22 complete player dataset](https://www.kaggle.com/datasets/stefanoleone992/fifa-22-complete-player-dataset)